### ISSUE TITLE
Avoid disk-full condition

### ISSUE THREAD

**Issue by jgfrm** [restic/restic#1153]
Restic fills the disk until the last byte.
If this happens, it is not possible to do anything:
- add more data (for the obvious reason)
- but more importantly, restore the data.

The way I solved it right now, is to create an 1GB file on my backup location, outside the directory where the restic backup is stored. In case of a full disk, I can remove this file and (hopefully) still restore the data.
But it would be nice if restic could prevent a full disk condition by itself.

**Comment by fd0** [comment]
You are right, thanks for raising this again. I doubt that we can implement a disk-full check for all backends (most "cloud services" will probably only return a generic error). Which one are you using right now?

**Comment by pvgoran** [comment]
Does restoration from the full-disk restic repository fail even if the restoration target is on another (non-full) disk?

**Comment by jgfrm** [comment]
I am using two actually:

- paid Strato storage via sftp

- free 1GB Stack from transip, which I mount locally via webdav

 

Van: Alexander Neumann [mailto:notifications@github.com] 
Verzonden: woensdag 9 augustus 2017 11:17
Aan: restic/restic <restic@noreply.github.com>
CC: jgfrm <frm@gordijn.org>; Author <author@noreply.github.com>
Onderwerp: Re: [restic/restic] Avoid disk-full condition (#1153)

 

You are right, thanks for raising this again. I doubt that we can implement a disk-full check for all backends (most "cloud services" will probably only return a generic error). Which one are you using right now?

**Comment by fd0** [comment]
Before restore, restic creates a lock file in the repo. If that fails it returns an error and exits. However, creating a lock file can be disabled by passing `--no-lock` to the `restore` operation, so it is possible to restore a backup from a repo which is on a disk filled up to the last byte.

**Comment by fd0** [comment]
@jgfrm by the way, what is your recommendation/expectation on what restic should do? Suppose you run `restic backup` and the FS mounted in via sftp has only 200MiB of storage left. What should it do exactly?

**Comment by jgfrm** [comment]
Ah, thank you very much.

Does the lock disabling also works to remove snapshots (because that would be the next step to free disk space)?

 

Van: Alexander Neumann [mailto:notifications@github.com] 
Verzonden: woensdag 9 augustus 2017 11:25
Aan: restic/restic <restic@noreply.github.com>
CC: jgfrm <frm@gordijn.org>; Author <author@noreply.github.com>
Onderwerp: Re: [restic/restic] Avoid disk-full condition (#1153)

 

Before restore, restic creates a lock file in the repo. If that fails it returns an error and exits. However, creating a lock file can be disabled by passing --no-lock to the restore operation, so it is possible to restore a backup from a repo which is on a disk filled up to the last byte.

**Comment by fd0** [comment]
No, removing snapshots (and more importantly: pruning/reorganising data) requires a lock file. That's a safety feature. In addition, the `prune` operation will always save all data that is still in use a second time in the repo into new files, and only at the very end remove unused files. So that operation (as it is now) will not work on a full disk.

I'm not really sure what we should do. By default I think restic should stay very conservative for all operations, which unfortunately means that it won't handle a full disk well.

I'm open to suggestions on what to do though because that situation will occur to users, and we need a way out.

**Comment by jgfrm** [comment]
If it is only the lock file issue, I would expect that the locking is still intact under a disk full condition.

(but then it should be possible to remove snapshots on a full disk).

I assume the lock file is just a temp file that is created.

Therefore, it is not really why things go wrong.

After all, the lock file is created (and therefore occupies diskspace) if a backup is made, and removed once the backup is finished.

So, after a backup, you should never have a disk full condition, since the lockfile takes diskspace and is removed afterwards.

But it seems that next time, the lock file can not be created.

 

Van: Alexander Neumann [mailto:notifications@github.com] 
Verzonden: woensdag 9 augustus 2017 11:27
Aan: restic/restic <restic@noreply.github.com>
CC: jgfrm <frm@gordijn.org>; Mention <mention@noreply.github.com>
Onderwerp: Re: [restic/restic] Avoid disk-full condition (#1153)

 

@jgfrm <https://github.com/jgfrm>  by the way, what is your recommendation/expectation on what restic should do? Suppose you run restic backup and the FS mounted in via sftp has only 200MiB of storage left. What should it do exactly?

**Comment by jgfrm** [comment]
Is there any heuristic you can use to know how much is space is needed to do the prune operation in case of a disk full condition?

If so, the following could be done:

- restic always create a file of the required space upon repository creation. This file is not removed.

- in case of a disk full condition, restic can work in a degradation mode only. In this mode, you can only free disk space in the repository. The degraded mode removes the file created during repository creation, so the removal/prune commands have enough space to their work.

- after removal, restic can brought into production modus again. This would imply that the file that occupies the required space is recreated.

 

The advantage is that this should work on remote repository.

- you have to detect disk full conditions (or leave this to the user, as he can quickly determine such a condistion manually).

- you have to store the state of repository in the repository itself (production|degraded_

- you may want to set the size of file that reserves space by a command line option, and/or restic can provide a meaningful suggestion.

- you need commands to put the repository in production/degraded state.

 

Van: Alexander Neumann [mailto:notifications@github.com] 
Verzonden: woensdag 9 augustus 2017 11:32
Aan: restic/restic <restic@noreply.github.com>
CC: jgfrm <frm@gordijn.org>; Mention <mention@noreply.github.com>
Onderwerp: Re: [restic/restic] Avoid disk-full condition (#1153)

 

No, removing snapshots (and more importantly: pruning/reorganising data) requires a lock file. That's a safety feature. In addition, the prune operation will always save all data that is still in use a second time in the repo into new files, and only at the very end remove unused files. So that operation (as it is now) will not work on a full disk.

I'm not really sure what we should do. By default I think restic should stay very conservative for all operations, which unfortunately means that it won't handle a full disk well.

I'm open to suggestions on what to do though because that situation will occur to users, and we need a way out.

**Comment by fd0** [comment]
This is a duplicate of #93, but since the discussion is here I'm closing #93.

**Comment by rawtaz** [comment]
This is also a duplicate of #1140, consolidate?

**Comment by wuast94** [comment]
is there anything new ? Im setting up rustic to OneDrive via rclone. I have just 1tb storage and I want something like last hours week monthly yearly etc. but at this point the time will come were my cloud storage is full. would be nice to set a limit in restic. 

for example "--max-space 200G" so rustic try to get the condition above done and when it will exceed the limit delete the last n snapshots perhaps enough space is free for the actual backup

**Comment by phuna** [comment]
I think @wuast94 's suggestion is a good option, I'm currently using Arq backup on Mac and it has the similar feature to prevent the full disk issue.

**Comment by Summerdave** [comment]
Adding functionality to remove old snapshots to save storage space to the `backup` command is a bad idea imho as i would not expect restic to _delete_ snapshots when backing up.
Still i think there has to be a way to specify a maximum repository size.
My proposal would be to include the functionality in the `forget` command. This way, one can run something like 
```
restic backup ...
restic forget --max-size 500G --keep-within 14d
restic prune
```
every week or so. However, this alters the semantics of the `forget` command quite a bit. When specifying a maximum repo size there, some of the other policies *will* be broken. One can think of multiple options there:

1. Delete snapshots oldest to latest (like @wuast94 proposed and many other tools handle it)
2. Break the other policies in some predefined order
3. Let the user decide which policy gets broken first
4. Delete snapshots biggest to smallest (in order to delete as few snapshots as possible)

Adding some `--strategy` parameter to `forget` would make this command even more confusing than it is right now, so i don't know if this is the right place at all.
The third option would be to introduce another command for this, however the functionality is so closely related to `forget` and `prune` that i think introducing another command would just cause confusion and be inconsistent. 
I don't think any option here is perfect but i like the second option (adding to `forget` with some king of strategy) the most.

**Comment by bleetsheep** [comment]
I managed to free 30GB on a full disk by removing some other files on the disk, now the only thing on my backup disk is the restic backup (so taking ~970/1000GB). Attempting to `forget` and `prune` some files (`prune` said it would free about 85GB), it just filled the disk again and threw a stack trace, but the latest backup looks intact. Reading the text above, it now makes sense to me:

> the `prune` operation will always save all data that is still in use a second time in the repo into new files, and only at the very end remove unused files. So that operation (as it is now) will not work on a full disk.

I do agree that being conservative is better than adding yolo subcommands, and I like some of the suggestions here like adding a flag to help restic manage the data in the way you want.

One simple workaround that should work on all server systems and local disks is limiting how much space it will try to free in each run, for example if you know that you have 30GB free, you could run `restic prune -r /mnt --available-space 30G` (or maybe `--available-space detect` for local disks). When `prune` finds that there is more than 30G to be freed, it will just stop rewriting packs (and whatever it does that takes more space) before that amount is reached. Next, the user can re-run the command to free more data (now doing 60GB in a single run, since we freed up 30) or restic could just do so by internally redoing the operation until it is complete. The latter is of course preferable, but I would be perfectly happy with an option that is more manual but way quicker to implement and might make its way into a release soon.

Or perhaps it could be even simpler and just catch the exception it currently throws upon a full disk in the `prune` operation, stop trying to rewrite packs at that point, and instead see which old packs it can remove because it already rewrote them. That might free some of the space (so long as it succeeded to rewrite at least one pack into a smaller version of itself) and the user can retry the operation a few times until it runs through.

This would still not work if there are zero bytes free, though, a situation which I now created. Removing files from the last `prune` attempt (whatever files it created last in the `data/` directory?) might be a way to solve this, but I guess resolving the situation vs. avoiding the situation is a different discussion.

**Comment by aawsome** [comment]
While not solving this issue, #2718 can improve the situation about pruning (note the `--max-repack-count` flag).

**Comment by zroug** [comment]
I like the idea of @Summerdave. My idea based on that is to add a flag like `--only-if-repo-size` to `forget`. Starting `forget` with this flag would execute the normal `forget` command only if the repository size is above the specified size. Otherwise the repository is not altered.

It doesn't change the semantics of `forget` much, is easy to understand and you can run `forget` with `--prune` after every backup, while the slow pruning operation is only done when necessary. Adding this flag to an existing `forget` command never removes more snapshots than without this flag, preventing the user from accidentally removing too many snapshots and adhering to the 'inclusive as possible' property of `forget`.

**Comment by romainreignier** [comment]
I was stuck with a back-up disk almost full (50 MB free) and prune was not working with the `no space left on device` message.
Seeing @aawsome comment, I have tried the latest git version and `prune` worked and I allowed me to free some space on the disk.
Thank you!

**Comment by rbo54** [comment]
This defect has been open almost four years. Are there any plans to fix this? I was backing up, hit disk full, and I lost everything in the repository. Most commands cause go stack traces. I tried removing some of the latest data, as well as the check option. Even snapshots would not show what was in repository.
It seems to me the right fix would be to keep an in memory transaction log of a backup, and if a disk full condition was hit, delete everything from the current backup, and then warn the user that more space needs to be made available.

**Comment by rawtaz** [comment]
@rbo54 If you lost everything in your repository, this is not due to restic. You cannot lose everything in your repository just because the disk goes full when you back up. The reason is simple: When backing up, restic only *adds* data to the repository (the only files it would delete are lock files which are very small and temporary files). So the data and snapshots you had prior to the current backup run, should still be there and be intact. If you "tried to remove some of the latest data" this might have been what caused your repository to become corrupt.

Calling it a defect is a bit overly accusive IMO. Every software out there is susceptible to the surrounding environment's resources running out, and it's certainly not a main concern of restic's to help you manage your disk space. You are the one responsible for making sure that you don't run out of disk space, and if you are unsure about it you should have monitoring in place.

That said, as is evident in this thread, this isn't an easy problem to solve for restic. There are so many different types of backends that it's not a simple matter of asking how much disk space there is left. Out of the suggestions in this issue, I think [the one by jgfrm](https://github.com/restic/restic/issues/1153#issuecomment-321206861) seems to make the most sense - that is, creating a space-reserving file in the repository, that can then be deleted if need be to free up some space for recovery actions.

However, even such a feature isn't easy to get right. How much space should be reserved is entirely dependent on how much data you store in the repository. For some, 1 GB might be enough, while for others this is nowhere near enough. Perhaps if we make a way by which the user can adjust the size of the reserved space as needed this would be useful though. It's probably one of the most practically viable solutions there is, it's simple and straight forward and would work with any backend.

**Comment by Vanav** [comment]
@rawtaz App should correctly recover from full disk condition, giving a way to purge old backups to free disk space. If the only way to recover is to delete full repository, it is a bug. There may be no reliable way to prevent full disk fill, or to enlarge physical disk.

**Comment by rbo54** [comment]
Fair enough. All I can say is that- as a new restic user- but one experienced in Linux and Go programming, it was not apparent to me how to get out of this situation. Is there a known procedure for how to do it? If so I'd be glad to recreate the problem and try the procedure. Is it necessary to have a non-restic reserve volume that can be removed in disk full situations? Obviously since I didn't know this problem existed, I did not have such reserve space. The entire volume was 100% full. I did not resort to manually deleting things until I could not think of anything else to try.

**Comment by MichaelEischer** [comment]
The simplest way at the moment is probably to mount an overlayfs over the repo to have some additional scratch space. Then use `forget` to remove some snapshots and afterwards prune with a small repack size, that is e.g. `prune  --max-repack-size 1G`. Note that this requires restic 0.12.0. Then merge back the upper layer of the overlayfs and repeat if necessary. It should be sufficient to have a scratch space of index size + max-repack-size + a bit extra.

The problem why a disk-full condition is so hard to resolve, is that it is in general fundamentally incompatible with the way how restic guarantees the repository integrity. restic ensures that no matter when it crashes or the backend becomes inaccessible the repository is still left in a sane state. As restic never modifies files, but only adds and removes them, this requires that for any change to the repository, restic first has to add new data before being able to safely delete no longer used data. This is obviously a problem when the disk is full.

Just keeping a little bit of reserved space won't be enough to automatically recover in most cases. For that we'd also have to restructure the `prune` command to work more incrementally, such that the scratch space required for `prune` is actually bounded.

**Comment by rbo54** [comment]
Thanks Michael. I tried several times to reproduce this (at least 3 times) this morning and could not. It was perfectly well behaved and there were no issues using check or prune. I am left not understanding what I saw yesterday.. Perhaps I triggered the error somehow but I don't know. One difference is that I was driving the backup from a program yesterday, feeding data into stdin. Another thing that may have been unusual is the program specified a -parent option on backup. Today I was doing backups from command line, I did try one with the -parent option and that made no difference. If I can reproduce it and gather any data I'll get back to you, but for now I cannot.

**Comment by bleetsheep** [comment]
> it is in general fundamentally incompatible with the way how restic guarantees the repository integrity. [...] restic first [adds] new data before being able to safely delete no longer used data

This also means that, if it runs into an error while backing things up, it hasn't yet deleted anything and it can safely undo additions to free up space without entering an 'insane' state. It doesn't need to modify files and gets you to a >0 bytes available state, which would then (usually) allow you to use `--max-repack-count` as mentioned above.

Undoing partial additions may not be what the user wants (even if that is an exceptionally rare case) and restic is often run non-interactively so we can't just ask. So I guess we would need to write a temp file with the list of files that were added in the current run? And remove that file right before it starts removing data, because as of then it is no longer safe to remove these new files. If restic runs into an error instead of getting to the delete part, one can revert to the last good state (with, presumably, some bytes available) by having restic do the Go equivalent of `<tmp/newly-added xargs rm`.

A poor man's way of doing this would be using `find -mmin -10 | xargs rm` to delete whatever you added in the last 10 minutes (or however long ago you started the backup), until restic has a proper solution implemented. That would work and be safe, assuming mtime is set correctly for each file, if I understand restic's repository integrity guarantee correctly?

Perhaps an overlayfs-based solution is easier, but I have very little experience with that. If we can build that into restic without too much trouble, I imagine it could be an --argument with the place to temporarily store files, that might be a nice one-step solution instead of this two-step scheme above (where you first have to delete incomplete data, then can rewrite packs).

**Comment by aawsome** [comment]
> This also means that, if it runs into an error while backing things up, it hasn't yet deleted anything and it can safely undo additions to free up space without entering an 'insane' state. It doesn't need to modify files and gets you to a >0 bytes available state, which would then (usually) allow you to use `--max-repack-count` as mentioned above.

This is not generally true. restic can potentially run in parallel by many systems. So while one backup fails due to a full disc a parallel backup could have already been finished **and** using already uploaded parts from the aborted backup.

`prune` already deletes pack files which are not indexed. If a backup run aborts because of a full disk, there will be likely some of these files and I think that a `prune --max-repack-size=0` has a quite high probability to succeed.

**Comment by bleetsheep** [comment]
Clearly I'm not familiar enough with the internals here, I thought restic did quite strict locking to prevent any parallel access at all. Either way, pruning with max repack 0 sounds like a good solution here then! I didn't realize it would work that way. Seems like that kind of resolves this ticket? Perhaps this workflow should still be documented somewhere (including the last-ditch overlayfs solution), but other than that, if the user tells restic to make a big backup and it all succeeds but uses the very last byte then restic did exactly what the user asked for.

Given that "it's not a simple matter of asking how much disk space there is left" (we can't reliably avoid it), this sounds as though restic currently implements all that it can realistically and reliably do to resolve the situation when it happens?

**Comment by MichaelEischer** [comment]
> Clearly I'm not familiar enough with the internals here, I thought restic did quite strict locking to prevent any parallel access at all.

The exclusive locking only applies to the check and prune commands. For backups there's no problem to run multiple backups in parallel.

> Given that "it's not a simple matter of asking how much disk space there is left" (we can't reliably avoid it), this sounds as though restic currently implements all that it can realistically and reliably do to resolve the situation when it happens?

It would definitely be possible to reduce the amount of free space necessary to be able to run `prune`. I assume that it could be possible with some optimizations to reach the range of a few (hundred) MBs.

**Comment by Baker0052** [comment]
My Workaround for that is the following:

Prune-Script

rm /mnt/restic/dummy.file
restic -r /mnt/restic -p /root/restic.p forget --keep-hourly=24 --keep-daily=30 --keep-monthly=6 --group-by path
restic -r /mnt/restic -p /root/restic.p prune --max-unused unlimited
restic -r /mnt/restic -p /root/restic.p prune --max-unused 0
restic -r /mnt/restic -p /root/restic.p check --read-data
fallocate -l 5G /mnt/restic/dummy.file

i keep an 5G Dummyfile on the repo storage.
Before every Prune the dummyfile gets deleted. Prune is deleting all packs which doesnt require (large) repacks (--max-unused unlimited), which frees a good amount of data and is fast.
then all unused packs get deleted (--max-unused 0).
After that the Dummyfile is getting created again.

**Comment by darioseidl** [comment]
I think, there's two issues being discussed here: 

1) restic backup is writing to the last available byte. When no disk is available a lot of commands fail (on Linux), but it's usually not too hard to recover by deleting some files.

I think it would be good if we could tell restic to keep at least some amount of disk space and abort the backup otherwise to avoid that situation. I understand that this might not be reliably possible for all backends, but it would be nice to have for local and sftp backends where it should be relatively straightforward. (For comparison, similar to what like the [Nextcloud client does](https://docs.nextcloud.com/desktop/2.6/advancedusage.html#low-disk-space)).

2) restic prune might fail if there is not enough disk space left, and that can make it difficult to get out of the above situation, if there are no other files that can be deleted. I don't know what could be done about that, only

> It would definitely be possible to reduce the amount of free space necessary to be able to run prune. I assume that it could be possible with some optimizations to reach the range of a few (hundred) MBs.

If that's possible, then it could make sense to tell restic to keep that many hundred MBs free disk space so that prune is always possible.

**Comment by darioseidl** [comment]
The dummy file is only suitable for a pure backup server, isn't it?
We have a local restic repo (in addition to a remote repo) on a server where other services are running too. (That might be a bit redundant, but that's what backups generally are). Anyway, I cannot use the dummy file approach because then the disk would still be seen as full by the other services. I guess, I could create a dummy file, run restic backup, forget, and prune, and then delete the file again?

**Comment by rawtaz** [comment]
It's suitable wherever you want to make sure that you can recover from a disk full condition. I understand now what you meant, you mean that restic should stop writing data when the amount of free disk space goes below a certain treshold. That's indeed different from a pure file occupying space that you can remove when shit hits the fan. The latter would work fine if you have a server where services don't share the same space though, e.g. if you use ZFS or whatever to give each service only a part of the disk. There's different ways to do this, but I see your point.

**Comment by bleetsheep** [comment]
Is using `prune --max-repack-size=0` after a failed backup not already a solution? There won't be a new snapshot that references the new packs, so it should resolve the problem to a point where you can iteratively reclaim space with `forget`ing and pruning.

To me it seems the ticket is a bit stuck, but that we could close it once that is documented as a solution so people know to use it. The necessary prune options are already merged through last year's pull request from aawsome above. Should I add those docs, or is there another solution that we are waiting for to be implemented?

**Comment by MichaelEischer** [comment]
Currently `prune --max-repack-size=1` would be necessary as `--max-repack-size=0` is currently ignored, which should be fixed, see #3591.

**Comment by smlx** [comment]
A far better general solution for this is to use disk quotas. Many filesystems implement this.
